src/vm/analysis/update_expressions_sorting/mod.rs

```diff
@@ -0,0 +1,216 @@
+use std::collections::{HashSet, HashMap};
+use std::iter::FromIterator;
+use vm::representations::{SymbolicExpression};
+use vm::representations::SymbolicExpressionType::{AtomValue, Atom, List};
+use vm::analysis::types::{ContractAnalysis, AnalysisPass};
+
+use super::AnalysisDatabase;
+use super::errors::{CheckResult, CheckError, CheckErrors};
+
+#[cfg(test)]
+mod tests;
+
+pub struct UpdateExpressionsSorting {
+    graph: Graph,
+    top_level_expressions_map: HashMap<String, TopLevelExpressionIndex>   
+}
+
+impl AnalysisPass for UpdateExpressionsSorting {
+
+    fn run_pass(contract_analysis: &mut ContractAnalysis, _analysis_db: &mut AnalysisDatabase) -> CheckResult<()> {
+        let mut command = UpdateExpressionsSorting::new();
+        command.run(contract_analysis)?;
+        Ok(())
+    }
+}
+
+impl <'a> UpdateExpressionsSorting {
+
+    fn new() -> Self {
+        Self { 
+            top_level_expressions_map: HashMap::new(),
+            graph: Graph::new()
+        }
+    }
+
+    pub fn run(&mut self, contract_analysis: &'a mut ContractAnalysis) -> CheckResult<()> {
+
+        let exprs = contract_analysis.expressions[..].to_vec();
+        for (expr_index, expr) in exprs.iter().enumerate() {
+            self.graph.add_node(expr_index)?;
+
+            match self.find_expression_definition(expr) {
+                Some((definition_name, atom_index, _)) => {
+                    let tle = TopLevelExpressionIndex { expr_index, atom_index };
+                    self.top_level_expressions_map.insert(definition_name, tle);
+                },
+                None => {}
+            }
+        }
+
+        for (expr_index, expr) in exprs.iter().enumerate() {
+            self.register_dependencies(&expr, expr_index)?;
+        }
+
+        let mut walker = GraphWalker::new();
+        let sorted_indexes = walker.get_sorted_dependencies(&self.graph)?;
+        
+        if let Some(deps) = walker.get_cycling_dependencies(&self.graph, &sorted_indexes) {
+            let deps_props: Vec<(String, u64, &SymbolicExpression)> = deps.iter().map(|i| {
+                let exp = &contract_analysis.expressions[*i];
+                self.find_expression_definition(&exp).unwrap()
+            }).collect();
+            let functions_names = deps_props.iter().map(|i| i.0.clone()).collect();
+            let exprs = deps_props.iter().map(|i| i.2.clone()).collect();
+
+            let mut error = CheckError::new(CheckErrors::CyclingDependencies(functions_names));
+            error.set_expressions(exprs);
+            return Err(error)
+        }
+
+        contract_analysis.top_level_expression_sorting = Some(sorted_indexes);
+        Ok(())
+    }
+
+    fn register_dependencies(&mut self, expr: &SymbolicExpression, tle_index: usize) -> CheckResult<bool> {
```

Comment:
I think you should add a code comment here that this function will _only_ work so long as variable shadowing is forbidden (which it is in Clarity). Otherwise, something like:

```
(define (bar x) (let (foo 1) (+ 1 x))
(define (foo x) (+ x 1))
```

Would incorrectly say that `bar` depended on `foo`.

Reply:
I've just pushed an iteration (and added tests), that should make this function work even if variable shadowing is allowed.